clang-tools-extra/clang-tidy/bugprone/NullCheckAfterDereferenceCheck.cpp

```diff
@@ -0,0 +1,125 @@
+//--------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "NullCheckAfterDereferenceCheck.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/AST/DeclCXX.h"
+#include "clang/AST/DeclTemplate.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+#include "clang/ASTMatchers/ASTMatchers.h"
+#include "clang/Analysis/CFG.h"
+#include "clang/Analysis/FlowSensitive/ControlFlowContext.h"
+#include "clang/Analysis/FlowSensitive/DataflowAnalysisContext.h"
+#include "clang/Analysis/FlowSensitive/DataflowEnvironment.h"
+#include "clang/Analysis/FlowSensitive/DataflowLattice.h"
+#include "clang/Analysis/FlowSensitive/Models/NullCheckAfterDereferenceModel.h"
+#include "clang/Analysis/FlowSensitive/WatchedLiteralsSolver.h"
+#include "clang/Basic/SourceLocation.h"
+#include "llvm/ADT/Any.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/Support/Error.h"
+#include <memory>
+#include <vector>
+
+namespace clang::tidy::bugprone {
+
+using ast_matchers::MatchFinder;
+using dataflow::NullCheckAfterDereferenceModel;
+using dataflow::ReverseNullDiagnoser;
+
+static constexpr llvm::StringLiteral FuncID("fun");
+
+static std::optional<ReverseNullDiagnoser::ResultType>
+analyzeFunction(const FunctionDecl &FuncDecl, ASTContext &ASTCtx) {
+  using dataflow::ControlFlowContext;
+  using dataflow::DataflowAnalysisState;
+  using llvm::Expected;
+
+  if (FuncDecl.getBody() == nullptr) {
+    return std::nullopt;
+  }
+
+  Expected<ControlFlowContext> Context =
+      ControlFlowContext::build(FuncDecl, *FuncDecl.getBody(), ASTCtx);
+  if (!Context)
+    return std::nullopt;
+
+  dataflow::DataflowAnalysisContext AnalysisContext(
+      std::make_unique<dataflow::WatchedLiteralsSolver>());
+  dataflow::Environment Env(AnalysisContext, FuncDecl);
+  NullCheckAfterDereferenceModel Analysis(ASTCtx);
+  Analysis.initializeFunctionParameters(*Context, Env);
+  ReverseNullDiagnoser Diagnoser;
+  ReverseNullDiagnoser::ResultType Diagnostics;
+  Expected<std::vector<std::optional<
+      DataflowAnalysisState<NullCheckAfterDereferenceModel::Lattice>>>>
+      BlockToOutputState = dataflow::runDataflowAnalysis(
+          *Context, Analysis, Env,
+          [&ASTCtx, &Diagnoser, &Diagnostics](
+              const CFGElement &Elt,
+              const DataflowAnalysisState<
+                  NullCheckAfterDereferenceModel::Lattice> &State) mutable {
+            auto EltDiagnostics = Diagnoser.diagnose(ASTCtx, &Elt, State.Env);
+            llvm::move(EltDiagnostics.first,
+                       std::back_inserter(Diagnostics.first));
+            llvm::move(EltDiagnostics.second,
+                       std::back_inserter(Diagnostics.second));
+          });
```

Comment:
The inner type of the `Expected`, if I read this correctly, is completely uninteresting, because your lambda takes care of populating the results. Without something better like `Expected<auto>` or `Expected<?>`, we could at least typedef it out as a "Detail" type and make the code more readable such that it has fewer kinds of parentheses everywhere.

```suggestion
  using LatticeState = DataflowAnalysisState<NullCheckAfterDereferenceModel::Lattice>;
  using DetailMaybeLatticeStates = std::vector<std::optional<LatticeState>>;
  
  auto DiagnosisImpl = [&ASTCtx, &Diagnoser, &Diagnostics](
              const CFGElement &Elt,
              const LatticeState &S) mutable -> void {
            auto EltDiagnostics = Diagnoser.diagnose(ASTCtx, &Elt, S.Env);
            llvm::move(EltDiagnostics.first,
                       std::back_inserter(Diagnostics.first));
            llvm::move(EltDiagnostics.second,
                       std::back_inserter(Diagnostics.second));
          };
          
  Expected<DetailMaybeLatticeStates>
      BlockToOutputState = dataflow::runDataflowAnalysis(
          *Context, Analysis, Env, DiagnosisImpl);
```